src/osd/scrubber/scrub_reservations.h

```diff
@@ -44,6 +47,20 @@ namespace Scrub {
  *  that have been acquired until that moment.
  *  (Why? because we have encountered instances where a reservation request was
  *  lost - either due to a bug or due to a network issue.)
+ *
+ * Keeping primary & replica in sync:
+ *
+ *  Reservation requests are unique in that they may time-out and cancelled
+ *  by the primary. A request that was delayed for too long and was
```

Comment:
Close.

```
Reservation requests may be canceled by the primary independently of the replica's response.
Depending on timing, a cancellation by the primary might or might not be processed by a replica
prior to sending a response (either rejection or success).  Thus, we associate each reservation
request with a nonce incremented with each reservation during an interval and drop any responses
that do not match our current nonce -- this behavior is occurs after rejecting any messages from
prior intervals so reusing nonces between intervals is not a problem.  Note that epoch would not
suffice as it's possible for this sequence to occur several times without a new map epoch.
```

In writing the above, I have a few questions:
- is it actually possible for the cancellation to order past the original request?  Or is it the case that we're actually guaranteed to see the response from the replica one way or the other?
- if we see a success response from the replica, we can assume that it saw or will see the cancellation, right?

Reply:
If the replica goes down, there will be a new message once it is detected (including if it goes down and comes back up) resulting in an interval change.  As long as the primary is still in the same interval, it may safely assume that every message will be delivered to a replica (any exception would be a serious bug).

Reply:
> so reusing nonces between intervals is not a problem

That comment made be aware of the fact that I was resetting the nonce too often. It is now moved to the
outer FSM state.